File: Data Structure/20255/Solutions/Find_Parent_Value_at_Given_Index_in_Max_Heap.py
class MaxHeap:
    def __init__(self):
        self.heap = []
    
    def insert(self, val):
        self.heap.append(val)
        self._heapify_up(len(self.heap) - 1)
    
    def _heapify_up(self, idx):
        parent = (idx - 1) // 2
        if idx > 0 and self.heap[idx] > self.heap[parent]:
            self.heap[idx], self.heap[parent] = self.heap[parent], self.heap[idx]
            self._heapify_up(parent)
    
    def findParent(self, index):
        if index == 0:
            return -1
        if index < 0 or index >= len(self.heap):
            return -1
        parent_idx = (index - 1) // 2
        return self.heap[parent_idx]
    
    def getHeap(self):
        return self.heap

def solve():
    n = int(input())
    arr = list(map(int, input().split()))
    index = int(input())
    heap = MaxHeap()
    for num in arr:
        heap.insert(num)
    result = heap.findParent(index)
    print(result)

if __name__ == '__main__':
    solve()

File: Data Structure/20255/Solutions/Find_Parent_Value_at_Given_Index_in_Max_Heap.js
class MaxHeap {
    constructor() {
        this.heap = [];
    }
    
    insert(val) {
        this.heap.push(val);
        this.heapifyUp(this.heap.length - 1);
    }
    
    heapifyUp(idx) {
        if (idx === 0) return;
        const parent = Math.floor((idx - 1) / 2);
        if (this.heap[idx] > this.heap[parent]) {
            [this.heap[idx], this.heap[parent]] = [this.heap[parent], this.heap[idx]];
            this.heapifyUp(parent);
        }
    }
    
    findParent(index) {
        if (index === 0) return -1;
        if (index < 0 || index >= this.heap.length) return -1;
        const parentIdx = Math.floor((index - 1) / 2);
        return this.heap[parentIdx];
    }
    
    getHeap() {
        return this.heap;
    }
}

function solve(input) {
    const lines = input.trim().split("\n");
    const n = parseInt(lines[0]);
    const arr = lines[1].split(" ").map(Number);
    const index = parseInt(lines[2]);
    const heap = new MaxHeap();
    for (let num of arr) heap.insert(num);
    const result = heap.findParent(index);
    console.log(result);
}

const fs = require('fs');
solve(fs.readFileSync(0, 'utf-8'));

